File: service/src/test/java/com/epam/esm/service/impl/OrderServiceImplTest.java
package com.epam.esm.service.impl;


import com.epam.esm.exception.EntityNotFoundException;
import com.epam.esm.model.entity.Order;
import com.epam.esm.repository.GiftCertificateRepository;
import com.epam.esm.repository.OrderRepository;
import com.epam.esm.repository.UserRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static com.epam.esm.service.impl.util.Constants.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OrderServiceImplTest {

    @Mock
    private OrderRepository orderRepository;

    @Mock
    private UserRepository userRepository;

    @Mock
    private GiftCertificateRepository giftCertificateRepository;

    @InjectMocks
    private OrderServiceImpl orderService;


    @Test
    void getAllByUserId() {
        List<Order> expected = Arrays.asList(FIRST_TEST_ORDER, SECOND_TEST_ORDER);
        when(orderRepository.getAllByUserId(TEST_ID, PAGE)).thenReturn(expected);
        when(userRepository.findById(TEST_ID)).thenReturn(Optional.of(FIRST_TEST_USER));

        List<Order> actual = orderService.getAllByUserId(TEST_ID, PAGE_NUM, PAGE_SIZE);

        assertEquals(expected, actual);
    }

    @Test
    void getAllByUserIdShouldThrowEntityNotFoundException() {
        when(userRepository.findById(NOT_EXIST_ID)).thenThrow(new EntityNotFoundException());
        assertThrows(EntityNotFoundException.class, () -> orderService.getAllByUserId(NOT_EXIST_ID, PAGE_NUM, PAGE_SIZE));
    }

    @Test
    void create() {
        when(userRepository.findById(TEST_ID)).thenReturn(Optional.of(FIRST_TEST_USER));
        when(giftCertificateRepository.findById(3L)).thenReturn(Optional.of(THIRD_TEST_GIFT_CERTIFICATE));
        when(orderRepository.save(THIRD_TEST_ORDER)).thenReturn(THIRD_TEST_ORDER);

        Order actual = orderService.create(TEST_ID, 3);
        assertEquals(THIRD_TEST_ORDER, actual);
    }

    @Test
    void createShouldThrowEntityNotFoundException() {
        when(userRepository.findById(NOT_EXIST_ID)).thenThrow(new EntityNotFoundException());

        assertThrows(EntityNotFoundException.class, () -> orderService.create(NOT_EXIST_ID, NOT_EXIST_ID));
    }

    @Test
    void getById() {
        when(orderRepository.findById(TEST_ID)).thenReturn(Optional.of(FIRST_TEST_ORDER));
        Order actual = orderService.getById(TEST_ID);
        assertEquals(FIRST_TEST_ORDER, actual);
    }

    @Test
    void getByIdShouldThrowEntityNotFoundException() {
        when(orderRepository.findById(NOT_EXIST_ID)).thenThrow(new EntityNotFoundException());
        assertThrows(EntityNotFoundException.class, () -> orderService.getById(NOT_EXIST_ID));
    }
}

File: service/src/test/java/com/epam/esm/service/impl/OrderServiceImplTest.kt
package com.epam.esm.service.impl

import com.epam.esm.repository.GiftCertificateRepository
import com.epam.esm.repository.OrderRepository
import com.epam.esm.repository.UserRepository
import com.epam.esm.service.impl.util.Constants.FIRST_TEST_ORDER
import com.epam.esm.service.impl.util.Constants.NOT_EXIST_ID
import com.epam.esm.service.impl.util.Constants.PAGE
import com.epam.esm.service.impl.util.Constants.PAGE_NUM
import com.epam.esm.service.impl.util.Constants.PAGE_SIZE
import com.epam.esm.service.impl.util.Constants.SECOND_TEST_ORDER
import com.epam.esm.service.impl.util.Constants.TEST_ID
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.test.StepVerifier
import org.mockito.Mockito.`when` as whenever

@ExtendWith(MockitoExtension::class)
class OrderServiceImplTest {

    @Mock
    private lateinit var orderRepository: OrderRepository

    @Mock
    private lateinit var userRepository: UserRepository

    @Mock
    private lateinit var giftCertificateRepository: GiftCertificateRepository

    @InjectMocks
    private lateinit var orderService: OrderServiceImpl

    @Test
    fun getAllByUserId() {
        val expected = listOf(FIRST_TEST_ORDER, SECOND_TEST_ORDER)

        whenever(orderRepository.findAllByUserId(TEST_ID, PAGE)).thenReturn(Flux.fromIterable(expected))

        val actual = orderService.getAllByUserId(TEST_ID, PAGE_NUM, PAGE_SIZE)

        StepVerifier.create(actual)
            .expectNextCount(expected.size.toLong())
            .verifyComplete()
    }

    @Test
    fun createShouldThrowEntityNotFoundException() {
        whenever(userRepository.findById(NOT_EXIST_ID)).thenReturn(Mono.empty())
        whenever(giftCertificateRepository.findById(TEST_ID)).thenReturn(Mono.empty())

        val actual = orderService.create(NOT_EXIST_ID, TEST_ID)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }

    @Test
    fun getById() {
        whenever(orderRepository.findById(TEST_ID)).thenReturn(Mono.just(FIRST_TEST_ORDER))

        val actual = orderService.getById(TEST_ID)

        StepVerifier.create(actual)
            .expectNext(FIRST_TEST_ORDER)
            .verifyComplete()
    }

    @Test
    fun getByIdShouldThrowEntityNotFoundException() {
        whenever(orderRepository.findById(NOT_EXIST_ID)).thenReturn(Mono.empty())

        val actual = orderService.getById(NOT_EXIST_ID)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }
}
